challenges/3/YOUR-CODE-GOES-HERE/hamidrezaNiakan/app/Http/Controllers/TravelSpotController.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Enums\TravelStatus;
+use App\Exceptions\InvalidTravelStatusForThisActionException;
+use App\Exceptions\ProtectedSpotException;
+use App\Exceptions\SpotAlreadyPassedException;
+use App\Http\Resources\TravelResource;
+use App\Models\Driver;
+use App\Models\Travel;
+use App\Models\TravelSpot;
+use Carbon\Carbon;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\DB;
+
+class TravelSpotController extends Controller {
+	public function arrived ( Travel $travel , TravelSpot $spot , Request $request ) {
+		DB::beginTransaction();
+		$user = $request->user();
+		if ( !Driver::isDriver($user) ) {
+			abort(403);
+		}
+		if ( $user->can('markAsArrived' , $spot) ) {
+			if ( $travel->driverHasArrivedToOrigin() ) {
+				return throw new SpotAlreadyPassedException();
+			}
+			if ( $travel->status->value == TravelStatus::CANCELLED->value ) {
+				throw new InvalidTravelStatusForThisActionException();
+			}
+			$origin_Spot = $travel->getOriginSpot();
+			$origin_Spot->arrived_at = Carbon::now();
+			$origin_Spot->save();
+			
+			return response()->json(TravelResource::make($travel));
+		}
+		DB::commit();
+	}
+	
+	public function store ( Travel $travel , Request $request ) {
+		$this->validate($request , [
```

Comment:
We already provide you `TravelSpotStoreRequest` to help you with validation.
This apply for all of other insert/update controllers too.